services/brig/src/Brig/API/Client.hs

```diff
@@ -151,52 +150,46 @@ rmClient u con clt pw =
         _ -> Data.reauthenticate u pw !>> ClientDataError . ClientReAuthError
       lift $ execDelete u (Just con) client
 
-claimPrekey :: MappedOrLocalId Id.U -> ClientId -> AppIO (Maybe ClientPrekey)
-claimPrekey u c = case u of
-  Local localUser ->
-    claimLocalPrekey localUser c
-  Mapped _ ->
-    -- FUTUREWORK(federation, #1272): claim key from other backend
-    pure Nothing
-
-claimLocalPrekey :: UserId -> ClientId -> AppIO (Maybe ClientPrekey)
-claimLocalPrekey u c = do
-  prekey <- Data.claimPrekey u c
-  case prekey of
-    Nothing -> noPrekeys u c >> return Nothing
-    pk@(Just _) -> return pk
-
-claimPrekeyBundle :: MappedOrLocalId Id.U -> AppIO PrekeyBundle
-claimPrekeyBundle = \case
-  Local localUser ->
-    claimLocalPrekeyBundle localUser
-  Mapped IdMapping {_imMappedId} ->
-    -- FUTUREWORK(federation, #1272): claim keys from other backend
-    pure $ PrekeyBundle (makeMappedIdOpaque _imMappedId) []
-
-claimLocalPrekeyBundle :: UserId -> AppIO PrekeyBundle
-claimLocalPrekeyBundle u = do
-  clients <- map clientId <$> Data.lookupClients u
-  PrekeyBundle (makeIdOpaque u) . catMaybes <$> mapM (Data.claimPrekey u) clients
+claimPrekey :: UserId -> Domain -> ClientId -> AppIO (Maybe ClientPrekey)
+claimPrekey u d c = do
+  isLocalDomain <- (d ==) <$> viewFederationDomain
+  if isLocalDomain
+    then claimLocalPrekey u c
+    else -- FUTUREWORK(federation, #1272): claim key from other backend
+      pure Nothing
+  where
+    claimLocalPrekey :: UserId -> ClientId -> AppIO (Maybe ClientPrekey)
+    claimLocalPrekey u' c' = do
+      prekey <- Data.claimPrekey u' c'
+      case prekey of
+        Nothing -> noPrekeys u' c' >> return Nothing
+        pk@(Just _) -> return pk
 
-claimMultiPrekeyBundles :: UserClients -> Handler (UserClientMap (Maybe Prekey))
-claimMultiPrekeyBundles (UserClients clientMap) = do
-  resolved <- lift . traverse (bitraverse resolveOpaqueUserId pure) $ Map.toList clientMap
-  let (localUsers, remoteUsers) = partitionEithers $ map localOrRemoteUser resolved
-  for_ (nonEmpty remoteUsers) $
-    throwStd . federationNotImplemented . fmap fst
-  -- FUTUREWORK(federation, #1272): claim keys from other backends, merge maps
-  lift $ UserClientMap . Map.mapKeys makeIdOpaque <$> claimLocalPrekeyBundles localUsers
+claimPrekeyBundle :: Domain -> UserId -> AppIO PrekeyBundle
+claimPrekeyBundle domain uid = do
+  isLocalDomain <- (domain ==) <$> viewFederationDomain
+  if isLocalDomain
+    then claimLocalPrekeyBundle uid
+    else -- FUTUREWORK(federation, #1272): claim keys from other backend
+      error "TODO FUTUREWORK(federation, #1272)"
```

Comment:
Let's please avoid hard errors like these for things not yet implemented. The previous code already had a `throwStd . federationNotImplemented`, which leads (currently) to a 403. We can discuss if 403 is the best, but that can be done at the definition of federationNotImplemented.

Plain 500s without good error messages are always worrying and hard to debug for both us (we see 500s in the metrics but have no logs here of what's happening) and for client developers.